File: src/meta.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
MIT License

Copyright (c) 2024 cauchy1988

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""


import copy
from typing import Dict, Optional


from parameters import Parameters
from node import Node
from request import Request
from vehicle import Vehicle


class Meta:
	def __init__(self, parameters: Parameters):
		# running parameters
		self.parameters = parameters
		
		# first_node_id -> {second_node_id -> distance between them}
		self.distances: Dict[int, Dict[int, float]] = {}
		# node_id -> Node
		self.nodes: Dict[int, Node] = {}
		# request_id -> Request
		self.requests: Dict[int, Request] = {}
		# vehicle_id -> Vehicle
		self.vehicles: Dict[int, Vehicle] = {}
		# vehicle_id -> {first_node_id -> {second_node_id -> run_time}}
		self.vehicle_run_between_nodes_time: Dict[int, Dict[int, Dict[int, float]]] = {}
	
	# this interface only use for problems with homogeneous fleet
	def add_one_same_vehicle(self, new_vehicle_id: Optional[int] = None) -> int :
		if len(self.vehicles) == 0:
			raise RuntimeError('vehicles is empty!')
		
		# Generate new vehicle ID if not provided
		if new_vehicle_id is None:
			max_vehicle_id = max(self.vehicles.keys())
			new_vehicle_id = max_vehicle_id + 1
		
		# Check if vehicle ID already exists
		if new_vehicle_id in self.vehicles:
			raise ValueError(f'Vehicle with ID {new_vehicle_id} already exists!')
		
		# Get reference vehicle (any vehicle since they should all be the same)
		reference_vehicle = next(iter(self.vehicles.values()))
		
		new_vehicle_start_node_id = max(self.nodes.keys()) + 1
		new_vehicle_end_node_id = new_vehicle_start_node_id + 1
		
		random_depot_node = self.nodes[reference_vehicle.start_node_id]
		self.nodes[new_vehicle_start_node_id] = Node(new_vehicle_start_node_id, random_depot_node.x, random_depot_node.y, random_depot_node.earliest_service_time, random_depot_node.latest_service_time, random_depot_node.service_time, random_depot_node.load)
		self.nodes[new_vehicle_end_node_id] = Node(new_vehicle_end_node_id, random_depot_node.x, random_depot_node.y, random_depot_node.earliest_service_time, random_depot_node.latest_service_time, random_depot_node.service_time, random_depot_node.load)

		# Add new vehicle to the Meta structure
		self.vehicles[new_vehicle_id] = Vehicle(
			new_vehicle_id,
			reference_vehicle.capacity,
			reference_vehicle.velocity,
			new_vehicle_start_node_id,
			new_vehicle_end_node_id
		)
		
		# Initialize vehicle_run_between_nodes_time for the new vehicle
		self.vehicle_run_between_nodes_time[new_vehicle_id] = {}
		for from_node_id in self.nodes.keys():
			self.vehicle_run_between_nodes_time[new_vehicle_id][from_node_id] = {}
			for to_node_id in self.nodes.keys():
				if from_node_id == to_node_id:
					self.vehicle_run_between_nodes_time[new_vehicle_id][from_node_id][to_node_id] = 0.0
				else:
					# Calculate travel time = distance / speed
					distance = self.distances.get(from_node_id, {}).get(to_node_id, 0.0)
					travel_time = distance / reference_vehicle.velocity if reference_vehicle.velocity > 0 else distance
					self.vehicle_run_between_nodes_time[new_vehicle_id][from_node_id][to_node_id] = travel_time
		
		# Add the new vehicle to all requests
		for one_request in self.requests.values():
			one_request.vehicle_set.add(new_vehicle_id)
			
		# Update distances
		random_depot_node_id = random_depot_node.node_id
		for from_node_id, to_node_dict in self.distances.items():
			to_node_dict[new_vehicle_start_node_id] = to_node_dict[random_depot_node_id]
			to_node_dict[new_vehicle_end_node_id] = to_node_dict[random_depot_node_id]
		self.distances[new_vehicle_start_node_id] = copy.deepcopy(self.distances[random_depot_node_id])
		self.distances[new_vehicle_start_node_id][new_vehicle_end_node_id] = 0.0
		self.distances[new_vehicle_end_node_id] = copy.deepcopy(self.distances[random_depot_node_id])
		self.distances[new_vehicle_end_node_id][new_vehicle_start_node_id] = 0.0
		
		for vehicle_id, time_dict in self.vehicle_run_between_nodes_time.items():
			for from_node_id, to_node_dict in time_dict.items():
				to_node_dict[new_vehicle_start_node_id] = to_node_dict[random_depot_node_id]
				to_node_dict[new_vehicle_end_node_id] = to_node_dict[random_depot_node_id]
			time_dict[new_vehicle_start_node_id] = copy.deepcopy(time_dict[random_depot_node_id])
			time_dict[new_vehicle_end_node_id] = copy.deepcopy(time_dict[random_depot_node_id])
			time_dict[new_vehicle_start_node_id][new_vehicle_end_node_id] = 0.0
			time_dict[new_vehicle_end_node_id][new_vehicle_start_node_id] = 0.0

		return new_vehicle_id
	
	def delete_vehicle(self, deleted_vehicle_id: int) -> bool:
		if deleted_vehicle_id not in self.vehicles:
			return False  # Return False to indicate nothing was deleted
		
		# Prevent deletion of the last vehicle
		if len(self.vehicles) <= 1:
			raise RuntimeError('Cannot delete the last vehicle! At least one vehicle must remain.')
		
		start_depot_node_id = self.vehicles[deleted_vehicle_id].start_node_id
		end_depot_node_id = self.vehicles[deleted_vehicle_id].end_node_id
		
		# Check if other vehicles are using these nodes - this should never happen if logic is correct
		start_node_in_use = any(veh.start_node_id == start_depot_node_id or veh.end_node_id == start_depot_node_id 
		                        for veh_id, veh in self.vehicles.items() if veh_id != deleted_vehicle_id)
		end_node_in_use = any(veh.start_node_id == end_depot_node_id or veh.end_node_id == end_depot_node_id 
		                      for veh_id, veh in self.vehicles.items() if veh_id != deleted_vehicle_id)
		
		# If nodes are shared, this is a bug - throw exception
		if start_node_in_use or end_node_in_use:
			raise RuntimeError(f"Vehicle {deleted_vehicle_id} has shared nodes with other vehicles: start_node_in_use={start_node_in_use}, end_node_in_use={end_node_in_use}")
		
		# Delete the vehicle from the Meta structure
		del self.vehicles[deleted_vehicle_id]
		
		# Delete from vehicle run time data
		if deleted_vehicle_id in self.vehicle_run_between_nodes_time:
			del self.vehicle_run_between_nodes_time[deleted_vehicle_id]
		
		# Delete nodes and all related data - they should definitely exist and not be shared
		del self.nodes[start_depot_node_id]
		del self.nodes[end_depot_node_id]
		
		# Delete from all time dictionaries
		for vehicle_id, time_dict in self.vehicle_run_between_nodes_time.items():
			for from_node_id, to_node_id_dict in time_dict.items():
				del to_node_id_dict[start_depot_node_id]
				del to_node_id_dict[end_depot_node_id]
			del time_dict[start_depot_node_id]
			del time_dict[end_depot_node_id]

		# Remove vehicle from all requests
		for one_request in self.requests.values():
			one_request.vehicle_set.discard(deleted_vehicle_id)  # discard() won't raise KeyError
		
		# Delete from distances
		for from_node_id, node_id_dict in self.distances.items():
			del node_id_dict[start_depot_node_id]
			del node_id_dict[end_depot_node_id]
		del self.distances[start_depot_node_id]
		del self.distances[end_depot_node_id]

		return True  # Return True to indicate successful deletion
	
	def get_max_distance(self) -> Optional[float]:
		if not self.distances:
			return None
		return float(max((v for d in self.distances.values() for v in d.values()), default=None))
	
	def max_vehicle_id(self) -> Optional[int]:
		if not self.vehicles:
			return None
		return max(self.vehicles.keys())
	
	def get_vehicle_count(self) -> int:
		"""Get the current number of vehicles."""
		return len(self.vehicles)
	

File: java-version/src/main/java/com/tc/pdwtw/model/Meta.java
package com.tc.pdwtw.model;

import com.tc.pdwtw.util.Parameters;
import java.util.*;

/**
 * MIT License
 * 
 * Copyright (c) 2024 cauchy1988
 * 
 * Meta class containing all problem data and configuration
 */
public class Meta {
    private Parameters parameters;
    private Map<Integer, Map<Integer, Double>> distances;
    private Map<Integer, Node> nodes;
    private Map<Integer, Request> requests;
    private Map<Integer, Vehicle> vehicles;
    private Map<Integer, Map<Integer, Map<Integer, Double>>> vehicleRunBetweenNodesTime;

    public Meta(Parameters parameters) {
        this.parameters = parameters;
        this.distances = new HashMap<>();
        this.nodes = new HashMap<>();
        this.requests = new HashMap<>();
        this.vehicles = new HashMap<>();
        this.vehicleRunBetweenNodesTime = new HashMap<>();
    }

    public Meta copy() {
        Meta newMeta = new Meta(this.parameters);
        
        // Deep copy distances
        for (Map.Entry<Integer, Map<Integer, Double>> entry : distances.entrySet()) {
            newMeta.distances.put(entry.getKey(), new HashMap<>(entry.getValue()));
        }
        
        // Deep copy nodes
        for (Map.Entry<Integer, Node> entry : nodes.entrySet()) {
            Node node = entry.getValue();
            newMeta.nodes.put(entry.getKey(), new Node(
                node.getNodeId(), node.getX(), node.getY(),
                node.getEarliestServiceTime(), node.getLatestServiceTime(),
                node.getServiceTime(), node.getLoad()
            ));
        }
        
        // Deep copy requests
        for (Map.Entry<Integer, Request> entry : requests.entrySet()) {
            Request req = entry.getValue();
            newMeta.requests.put(entry.getKey(), new Request(
                req.getIdentity(), req.getPickNodeId(), req.getDeliveryNodeId(),
                req.getRequireCapacity(), new HashSet<>(req.getVehicleSet())
            ));
        }
        
        // Deep copy vehicles
        for (Map.Entry<Integer, Vehicle> entry : vehicles.entrySet()) {
            Vehicle veh = entry.getValue();
            newMeta.vehicles.put(entry.getKey(), new Vehicle(
                veh.getIdentity(), veh.getCapacity(), veh.getVelocity(),
                veh.getStartNodeId(), veh.getEndNodeId()
            ));
        }
        
        // Deep copy vehicle run times
        for (Map.Entry<Integer, Map<Integer, Map<Integer, Double>>> vehEntry : vehicleRunBetweenNodesTime.entrySet()) {
            Map<Integer, Map<Integer, Double>> vehTimes = new HashMap<>();
            for (Map.Entry<Integer, Map<Integer, Double>> fromEntry : vehEntry.getValue().entrySet()) {
                vehTimes.put(fromEntry.getKey(), new HashMap<>(fromEntry.getValue()));
            }
            newMeta.vehicleRunBetweenNodesTime.put(vehEntry.getKey(), vehTimes);
        }
        
        return newMeta;
    }

    public int addOneSameVehicle(Integer newVehicleId) {
        if (vehicles.isEmpty()) {
            throw new RuntimeException("vehicles is empty!");
        }

        if (newVehicleId == null) {
            int maxVehicleId = Collections.max(vehicles.keySet());
            newVehicleId = maxVehicleId + 1;
        }

        if (vehicles.containsKey(newVehicleId)) {
            throw new IllegalArgumentException("Vehicle with ID " + newVehicleId + " already exists!");
        }

        Vehicle referenceVehicle = vehicles.values().iterator().next();
        
        int newVehicleStartNodeId = Collections.max(nodes.keySet()) + 1;
        int newVehicleEndNodeId = newVehicleStartNodeId + 1;

        Node randomDepotNode = nodes.get(referenceVehicle.getStartNodeId());
        nodes.put(newVehicleStartNodeId, new Node(
            newVehicleStartNodeId, randomDepotNode.getX(), randomDepotNode.getY(),
            randomDepotNode.getEarliestServiceTime(), randomDepotNode.getLatestServiceTime(),
            randomDepotNode.getServiceTime(), randomDepotNode.getLoad()
        ));
        nodes.put(newVehicleEndNodeId, new Node(
            newVehicleEndNodeId, randomDepotNode.getX(), randomDepotNode.getY(),
            randomDepotNode.getEarliestServiceTime(), randomDepotNode.getLatestServiceTime(),
            randomDepotNode.getServiceTime(), randomDepotNode.getLoad()
        ));

        vehicles.put(newVehicleId, new Vehicle(
            newVehicleId, referenceVehicle.getCapacity(), referenceVehicle.getVelocity(),
            newVehicleStartNodeId, newVehicleEndNodeId
        ));

        // Initialize vehicle run times
        vehicleRunBetweenNodesTime.put(newVehicleId, new HashMap<>());
        for (Integer fromNodeId : nodes.keySet()) {
            vehicleRunBetweenNodesTime.get(newVehicleId).put(fromNodeId, new HashMap<>());
            for (Integer toNodeId : nodes.keySet()) {
                if (fromNodeId.equals(toNodeId)) {
                    vehicleRunBetweenNodesTime.get(newVehicleId).get(fromNodeId).put(toNodeId, 0.0);
                } else {
                    double distance = distances.getOrDefault(fromNodeId, new HashMap<>()).getOrDefault(toNodeId, 0.0);
                    double travelTime = referenceVehicle.getVelocity() > 0 ? distance / referenceVehicle.getVelocity() : distance;
                    vehicleRunBetweenNodesTime.get(newVehicleId).get(fromNodeId).put(toNodeId, travelTime);
                }
            }
        }

        // Add vehicle to all requests
        for (Request request : requests.values()) {
            request.getVehicleSet().add(newVehicleId);
        }

        // Update distances
        int randomDepotNodeId = randomDepotNode.getNodeId();
        for (Map<Integer, Double> toNodeDict : distances.values()) {
            toNodeDict.put(newVehicleStartNodeId, toNodeDict.get(randomDepotNodeId));
            toNodeDict.put(newVehicleEndNodeId, toNodeDict.get(randomDepotNodeId));
        }
        
        distances.put(newVehicleStartNodeId, new HashMap<>(distances.get(randomDepotNodeId)));
        distances.get(newVehicleStartNodeId).put(newVehicleEndNodeId, 0.0);
        distances.put(newVehicleEndNodeId, new HashMap<>(distances.get(randomDepotNodeId)));
        distances.get(newVehicleEndNodeId).put(newVehicleStartNodeId, 0.0);

        // Update vehicle run times for all vehicles
        for (Map<Integer, Map<Integer, Double>> timeDict : vehicleRunBetweenNodesTime.values()) {
            for (Map<Integer, Double> toNodeDict : timeDict.values()) {
                toNodeDict.put(newVehicleStartNodeId, toNodeDict.get(randomDepotNodeId));
                toNodeDict.put(newVehicleEndNodeId, toNodeDict.get(randomDepotNodeId));
            }
            timeDict.put(newVehicleStartNodeId, new HashMap<>(timeDict.get(randomDepotNodeId)));
            timeDict.put(newVehicleEndNodeId, new HashMap<>(timeDict.get(randomDepotNodeId)));
            timeDict.get(newVehicleStartNodeId).put(newVehicleEndNodeId, 0.0);
            timeDict.get(newVehicleEndNodeId).put(newVehicleStartNodeId, 0.0);
        }

        return newVehicleId;
    }

    public boolean deleteVehicle(int deletedVehicleId) {
        if (!vehicles.containsKey(deletedVehicleId)) {
            return false;
        }

        if (vehicles.size() <= 1) {
            throw new RuntimeException("Cannot delete the last vehicle! At least one vehicle must remain.");
        }

        Vehicle deletedVehicle = vehicles.get(deletedVehicleId);
        int startDepotNodeId = deletedVehicle.getStartNodeId();
        int endDepotNodeId = deletedVehicle.getEndNodeId();

        // Check if nodes are shared (should not happen)
        boolean startNodeInUse = vehicles.values().stream()
            .filter(v -> v.getIdentity() != deletedVehicleId)
            .anyMatch(v -> v.getStartNodeId() == startDepotNodeId || v.getEndNodeId() == startDepotNodeId);
        boolean endNodeInUse = vehicles.values().stream()
            .filter(v -> v.getIdentity() != deletedVehicleId)
            .anyMatch(v -> v.getStartNodeId() == endDepotNodeId || v.getEndNodeId() == endDepotNodeId);

        if (startNodeInUse || endNodeInUse) {
            throw new RuntimeException("Vehicle " + deletedVehicleId + " has shared nodes with other vehicles");
        }

        // Delete vehicle
        vehicles.remove(deletedVehicleId);
        vehicleRunBetweenNodesTime.remove(deletedVehicleId);

        // Delete nodes
        nodes.remove(startDepotNodeId);
        nodes.remove(endDepotNodeId);

        // Update time dictionaries
        for (Map<Integer, Map<Integer, Double>> timeDict : vehicleRunBetweenNodesTime.values()) {
            for (Map<Integer, Double> toNodeDict : timeDict.values()) {
                toNodeDict.remove(startDepotNodeId);
                toNodeDict.remove(endDepotNodeId);
            }
            timeDict.remove(startDepotNodeId);
            timeDict.remove(endDepotNodeId);
        }

        // Remove vehicle from all requests
        for (Request request : requests.values()) {
            request.getVehicleSet().remove(deletedVehicleId);
        }

        // Update distances
        for (Map<Integer, Double> nodeDict : distances.values()) {
            nodeDict.remove(startDepotNodeId);
            nodeDict.remove(endDepotNodeId);
        }
        distances.remove(startDepotNodeId);
        distances.remove(endDepotNodeId);

        return true;
    }

    public Double getMaxDistance() {
        if (distances.isEmpty()) {
            return null;
        }
        return distances.values().stream()
            .flatMap(map -> map.values().stream())
            .max(Double::compareTo)
            .orElse(null);
    }

    public Integer maxVehicleId() {
        return vehicles.isEmpty() ? null : Collections.max(vehicles.keySet());
    }

    public int getVehicleCount() {
        return vehicles.size();
    }

    // Getters and Setters
    public Parameters getParameters() { return parameters; }
    public void setParameters(Parameters parameters) { this.parameters = parameters; }
    
    public Map<Integer, Map<Integer, Double>> getDistances() { return distances; }
    public void setDistances(Map<Integer, Map<Integer, Double>> distances) { this.distances = distances; }
    
    public Map<Integer, Node> getNodes() { return nodes; }
    public void setNodes(Map<Integer, Node> nodes) { this.nodes = nodes; }
    
    public Map<Integer, Request> getRequests() { return requests; }
    public void setRequests(Map<Integer, Request> requests) { this.requests = requests; }
    
    public Map<Integer, Vehicle> getVehicles() { return vehicles; }
    public void setVehicles(Map<Integer, Vehicle> vehicles) { this.vehicles = vehicles; }
    
    public Map<Integer, Map<Integer, Map<Integer, Double>>> getVehicleRunBetweenNodesTime() { return vehicleRunBetweenNodesTime; }
    public void setVehicleRunBetweenNodesTime(Map<Integer, Map<Integer, Map<Integer, Double>>> vehicleRunBetweenNodesTime) { 
        this.vehicleRunBetweenNodesTime = vehicleRunBetweenNodesTime; 
    }
}